winlogbeat/sys/wineventlog/wineventlog_windows.go

```diff
@@ -403,35 +403,35 @@ func FormatEventString(
 	}
 
 	var bufferPtr *byte
-	if renderBuf != nil {
+	if len(renderBuf) > 0 {
 		bufferPtr = &renderBuf[0]
 	}
 
 	// EvtFormatMessage operates with WCHAR buffer, assuming the size of the buffer in characters.
 	// https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage
-	var bufferNeeded uint32
+	var bufferUsed uint32
 	bufferSize := uint32(len(renderBuf) / 2)
 
-	err := _EvtFormatMessage(ph, eventHandle, 0, 0, nil, messageFlag, bufferSize, bufferPtr, &bufferNeeded)
-	if err != nil && err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // This is an errno.
+	err := _EvtFormatMessage(ph, eventHandle, 0, 0, nil, messageFlag, bufferSize, bufferPtr, &bufferUsed)
+	if err != nil && !errors.Is(err, windows.ERROR_INSUFFICIENT_BUFFER) {
 		return fmt.Errorf("failed in EvtFormatMessage: %w", err)
 	} else if err == nil {
-		// Windows API returns a null terminated WCHAR C-style string in the buffer. bufferNeeded applies
-		// only when ERROR_INSUFFICIENT_BUFFER is returned. Luckily the UTF16ToUTF8Bytes/UTF16ToString
-		// functions stop at null termination. Note, as signaled in a comment at the end of this function,
-		// this behavior is bad for EvtFormatMessageKeyword as then the API returns a list of null terminated
-		// strings in the buffer (it's fine for now as we don't use this parameter value).
-		return common.UTF16ToUTF8Bytes(renderBuf, out)
+		// bufferUsed indicates the size used internally to render the message. When called with nil buffer
+		// EvtFormatMessage returns ERROR_INSUFFICIENT_BUFFER, but otherwise succeeds copying only up to
+		// bufferSize to our buffer, truncating the message if our buffer was too small.
+		if bufferUsed <= bufferSize {
```

Comment:
When truncation occurs, doesn't it return exactly the input buffer size?  That's why we were seeing exactly 8192 chars of truncated output with a 16KB buffer.  If that's right, then this should be a strict less-than.

```suggestion
		if bufferUsed < bufferSize {
```

Reply:
I can't confirm that "it return exactly the input buffer size". I remember getting that impression when making the first look at why it's 8KB discovery, but debugging doesn't confirm this. I was setting a small buffer `renberBuf = make([]byte, 100)` just before the call to confirm and I always got the right `bufferUsed`.

Reply:
> I was setting a small buffer `renberBuf = make([]byte, 100)` just before the call to confirm and I always got the right `bufferUsed`.

What was the right `bufferUsed` for that scenario?

Reply:
My reading of what you said here contradicts the PR description.

> Windows API documentation is not very clear about it but the APIs return ERROR_INSUFFICIENT_BUFFER only when used with nil buffer, otherwise they happily succeed copying as much as the given buffer allows (truncating message)

Should one be updated?

Reply:
> // bufferUsed indicates the size used internally to render the message. When called with nil buffer
		// EvtFormatMessage returns ERROR_INSUFFICIENT_BUFFER, but otherwise succeeds copying only up to
		// bufferSize to our buffer, truncating the message if our buffer was too small.
		
What should be updated here or in the description? Both tell the same story. Should I mention the `bufferUsed` in description?